content/translation.js

```diff
@@ -0,0 +1,143 @@
+    before.push(macroArgs);
+    macros.set(macroName, before);
+  }
+
+  if (cacheKey) {
+    cacheKSMacros.set(cacheKey, macros);
+  }
+  return macros;
+}
+
+const cacheContentFeatures = new LRU({ max: 2000 });
+
+function getContentFeatures(rawContent, cacheKey = null) {
+  if (cacheKey && cacheContentFeatures.has(cacheKey)) {
+    return cacheContentFeatures.get(cacheKey);
+  }
+  const features = new Map();
+  const $ = cheerio.load(rawContent);
+  $("h2, h3, p").each((i, element) => {
+    const { tagName } = element;
+    if (!features.has(tagName)) {
+      features.set(tagName, new Set());
+    }
+    const text = $(element).text().trim();
+    if (text.startsWith("{{") && text.endsWith("}}")) {
+      // E.g. `{{NextMenu("Learn/Forms/How_to_structure_a_web_form", "Learn/Forms")}}`
+      // These are too hard to compare and just too weird. So for now, just skip them.
+      return;
+    }
+    if (text) {
+      features.get(tagName).add(text);
+    }
+  });
+  if (cacheKey) {
+    cacheContentFeatures.set(cacheKey, features);
+  }
+  return features;
+}
+
+const IMPORTANT_MACROS = new Map(
+  [
+    "Compat",
+    "EmbedInteractiveExample",
+    "EmbedLiveSample",
+    "EmbedGHLiveSample",
+    "CSSSyntax",
+    // XXX List all the important sidebar macros??
+  ].map((name) => [name.toLowerCase(), name])
+);
+
+function* getTranslationDifferences(englishDocument, translatedDocument) {
+  // Compare key KS macros presence
+  const translatedMacros = getKSMacros(translatedDocument.rawContent);
+  const englishMacros = getKSMacros(
+    englishDocument.rawContent,
+    englishDocument.url
+  );
+  for (const [macro, macroName] of IMPORTANT_MACROS) {
+    const macrosEnglish = englishMacros.get(macro) || [];
+    // const countEnglish = macrosEnglish.length;
+    const macrosTranslation = translatedMacros.get(macro) || [];
+    // const countTranslation = macrosTranslation.length;
+    if (macrosEnglish.length !== macrosTranslation.length) {
+      // Independent of the arguments, the *presence* of important macros is different.
+      const fullExplanation = [];
+      yield {
+        type: "macro",
+        name: macroName,
+        // XXX Might be nice to make this a bit more "human".
+        // For example, a lot of times this becomes:
+        //    EmbedInteractiveExample macro: en-US has 1, zh-CN has 0
+        // which is a bit cryptic.
+        explanation: `en-US has ${macrosEnglish.length}, ${translatedDocument.metadata.locale} has ${macrosTranslation.length}`,
+        fullExplanation,
+      };
+    } else if (macrosEnglish.length > 0) {
+      // XXX compare arguments?
+    }
+  }
+
+  // Compare possibly untranslated remaining English paragraphs, h2, and h3s.
+  const translatedContentFeatures = getContentFeatures(
+    translatedDocument.rawContent
+  );
+  const englishContentFeatures = getContentFeatures(
+    englishDocument.rawContent,
+    englishDocument.url
+  );
+  for (const tagName of ["h2", "h3", "p"]) {
```

Comment:
I like the idea in general. There's some false positives where `h2`'s won't be translated like the work "DOM" or "JavaScript". For `p`s it will happen all the time that it's an outdated English version in the translated article and these won't get flagged.

Reply:
Yeah. And this speaks to your point above. With that "uncertainty" and the fact that there is a greater-than-zero chance of false positives, it really ought to be *warnings*. Not flaws. 
Flaws generally mean it's expected to need work; just a matter of *when* you deal with it. 

In all fairness, flaws is a nice "framework" because you can do things like setting levels, and there's a good place for rendering it in the Toolbar. 

But some things are flaws which you genuinely should address. 

Perhaps what we can do is refactor how it all works. There could be a separate function that generates structured translation-differences-warnings. E.g. 
```
const translationDifferenceWarnings = getTranslationDIfferences(thisDoc.URL);
addWarnings("translation_differences", thisDoc, translationDifferenceWarnings);
if (translationDifferencesWarnings.macros) {
  addFlaws("translation_differences", thisDoc, translationDifferencesWarnings.macros)
}
```
(And the Toolbar React component can notice if the doc has a `.warnings` and display that adjacent to flaws)

The other thing I realize is that right now, my code sits in `build/flaws.js` but it actually doesn't analyze differences in the *built HTML* (after KS rendering and other cheerio mutations). 
So if it lived elsewhere, it could be purely related to comparing the *source* HTML with the en-us source HTML. 
And if it lived outside the rendering, you could write a query (that is sufficiently fast) that compares the differences so you can list all the translated sources that have differences. 

Having said that, at the same time, if you analyze the *built* HTML you could technically compare a en-us `index.md` with a fr `index.html`.